### ISSUE TITLE
Simplify `Option(Union(Option, ...))`

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#1337]
## Simplifying Union + Option

Per a conversation with @jpivarski, we require that Awkward functions produce "simplified" layouts as-policy. These simplification rules ultimately mean that layout visitors need less local information in order to reason about the structure.

We simplify nested `is_OptionType`s, `is_IndexedType`s, and their cross combinations, but that leaves some other candidates:
1. `Option(Union(Option, ...))`
2. `Indexed(Union(...))`

### `Option(Union(Option ,...))`
PR #1249 handles (1) for `ak.is_none`, but it might be better to solve this for _all_ functions by extending the simplification rules such that options simplify with unions. 

@jpivarski wrote in #1335 that we might benefit from choosing a canonical form for (1).  

To my mind, the simplest conceptual approach is to move the option into the union à la `Union(Option(X), Option(Y), Option(Z))`. This rule would read:

> If the content of an option type is a union **AND** the union has any optional contents **THEN** move the option into the union

This would be the least restrictive solution, as we could have:
```
Option[Byte](Union(X, Y, Option[?](Z))) → Union(Option[Byte](X), Option[Byte](Y), Option[?](Z))
Option[Bit](Union(X, Y, Option[?](Z))) → Union(Option[Bit](X), Option[Bit](Y), Option[?](Z))
Option[Indexed](Union(X, Y, Option[?](Z))) → Union(Option[Indexed](X), Option[Indexed](Y), Option[Indexed](Z))
```
It would also, I think, be the simplest to implement.

@jpivarski suggested
```
Option(Union(X, Y, Option(Z))) → ByteMaskedArray(UnionArray([X, Y, Z]))
```

i.e. lifting the option type out of the union. This would require a little more thinking in my opinion, namely:
- Cases where the internal `Option(Z)` is an `IndexedOptionArray` would require additional work to preserve the length of `Option(Z)` after dropping the option (e.g. mapping to random elements where `Option(Z).index == -1`.
- OR any union member with `IndexedOptionArray` would force the outer option to also be indexed.

The TL;DR for this case is that the `UnionArray` needs to map the `-1` indices of `Option(Z)` to *something* in order for the outer `ByteMaskedArray` to mask them.

### `Indexed(Union(...))`
The simplification function to handle this case would be better named `simplify_indexedtype`, however we already handle other scenarios with `simplify_optiontype`, so I think it's valid.

A `UnionArray` can be re-indexed by modifying its `tags` and `index` parameters. This means that the outer `Index` is redundant.

**Comment by agoose77** [comment]
This is now true; `indexed[union[...]]` is not permitted (save for categoricals), and `option[union`[...]]` is not permitted.
